File: java/new_ds/SkipList.java
package new_ds;

import java.util.Random;

/**
 * Skip List — a probabilistic data structure that allows O(log n) average-case
 * search, insertion, and deletion by maintaining multiple linked-list "levels".
 *
 * <p>Each node at level {@code i} is also present at level {@code i-1}.
 * A coin flip (p = 0.5) determines how many levels a new node is promoted to.
 *
 * <pre>
 * Time Complexity (average):
 *   insert   – O(log n)
 *   search   – O(log n)
 *   delete   – O(log n)
 *
 * Time Complexity (worst):
 *   insert   – O(n)
 *   search   – O(n)
 *   delete   – O(n)
 *
 * Space Complexity: O(n log n) expected
 * </pre>
 *
 * @param <T> the type of values stored in the list
 */
public class SkipList<T> {

    // -------------------------------------------------------------------------
    // Constants
    // -------------------------------------------------------------------------

    /** Maximum number of levels in the skip list. */
    private static final int MAX_LEVEL = 16;

    /** Probability of promoting a node to the next level. */
    private static final double P = 0.5;

    // -------------------------------------------------------------------------
    // Inner node class
    // -------------------------------------------------------------------------

    /**
     * A node in the skip list.
     *
     * <p>Each node stores a key-value pair and an array of {@code forward}
     * pointers — one per level at which this node participates.
     * {@code forward[0]} is the ordinary next pointer (level 0).
     *
     * @param <V> value type
     */
    @SuppressWarnings("unchecked")
    public static class SkipListNode<V> {

        /** Integer search key. */
        final int key;

        /** User-supplied value; may be {@code null} for sentinel nodes. */
        V value;

        /**
         * Forward pointers indexed by level.
         * {@code forward[i]} points to the next node at level {@code i}.
         */
        final SkipListNode<V>[] forward;

        /** Number of levels at which this node participates (its "height"). */
        final int level;

        /**
         * Constructs a node with the given key, value, and level.
         *
         * @param key   the integer key
         * @param value the associated value
         * @param level the height of this node (number of forward pointers)
         */
        SkipListNode(int key, V value, int level) {
            this.key     = key;
            this.value   = value;
            this.level   = level;
            this.forward = new SkipListNode[level + 1];
        }

        @Override
        public String toString() {
            return "(" + key + ":" + value + ")";
        }
    }

    // -------------------------------------------------------------------------
    // Fields
    // -------------------------------------------------------------------------

    /** Sentinel header node (key = Integer.MIN_VALUE, not visible to callers). */
    private final SkipListNode<T> head;

    /** Current highest level in use (0-indexed). */
    private int currentLevel;

    /** Number of key-value pairs stored. */
    private int size;

    /** Source of randomness for level generation. */
    private final Random random;

    // -------------------------------------------------------------------------
    // Constructor
    // -------------------------------------------------------------------------

    /**
     * Creates an empty skip list.
     */
    public SkipList() {
        head         = new SkipListNode<>(Integer.MIN_VALUE, null, MAX_LEVEL);
        currentLevel = 0;
        size         = 0;
        random       = new Random();
    }

    // -------------------------------------------------------------------------
    // Private helpers
    // -------------------------------------------------------------------------

    /**
     * Generates a random level for a new node using geometric distribution.
     *
     * <p>Starting at level 1, each successive level is included with
     * probability {@link #P} until {@link #MAX_LEVEL} is reached.
     *
     * @return a level in [1, MAX_LEVEL]
     */
    private int randomLevel() {
        int level = 1;
        while (random.nextDouble() < P && level < MAX_LEVEL) {
            level++;
        }
        return level;
    }

    /**
     * Builds the {@code update} array used by insert and delete.
     *
     * <p>{@code update[i]} is the rightmost node at level {@code i} whose key
     * is strictly less than {@code key}.
     *
     * @param key the target key
     * @return update array of length MAX_LEVEL + 1
     */
    @SuppressWarnings("unchecked")
    private SkipListNode<T>[] buildUpdate(int key) {
        SkipListNode<T>[] update = new SkipListNode[MAX_LEVEL + 1];
        SkipListNode<T>   curr   = head;

        for (int i = currentLevel; i >= 0; i--) {
            while (curr.forward[i] != null && curr.forward[i].key < key) {
                curr = curr.forward[i];
            }
            update[i] = curr;
        }
        return update;
    }

    // -------------------------------------------------------------------------
    // Public API
    // -------------------------------------------------------------------------

    /**
     * Inserts or updates a key-value pair.
     *
     * <p>If {@code key} already exists the value is overwritten.
     * Otherwise a new node is created at a randomly chosen level.
     *
     * <p>Time: O(log n) average &nbsp;|&nbsp; Space: O(log n) average (new node)
     *
     * @param key   the integer key
     * @param value the value to associate with {@code key}
     */
    public void insert(int key, T value) {
        SkipListNode<T>[] update = buildUpdate(key);

        // The candidate node immediately to the right at level 0.
        SkipListNode<T> curr = update[0].forward[0];

        // Key already exists — update in place.
        if (curr != null && curr.key == key) {
            curr.value = value;
            return;
        }

        int newLevel = randomLevel();

        // If new node is taller than anything seen so far, extend update array.
        if (newLevel > currentLevel) {
            for (int i = currentLevel + 1; i <= newLevel; i++) {
                update[i] = head;
            }
            currentLevel = newLevel;
        }

        SkipListNode<T> newNode = new SkipListNode<>(key, value, newLevel);

        // Splice new node into each level.
        for (int i = 0; i <= newLevel; i++) {
            newNode.forward[i]   = update[i].forward[i];
            update[i].forward[i] = newNode;
        }

        size++;
    }

    /**
     * Searches for a key and returns its associated value, or {@code null}
     * if the key is not present.
     *
     * <p>Time: O(log n) average
     *
     * @param key the integer key to look up
     * @return the associated value, or {@code null}
     */
    public T search(int key) {
        SkipListNode<T> curr = head;

        for (int i = currentLevel; i >= 0; i--) {
            while (curr.forward[i] != null && curr.forward[i].key < key) {
                curr = curr.forward[i];
            }
        }

        curr = curr.forward[0];
        if (curr != null && curr.key == key) {
            return curr.value;
        }
        return null;
    }

    /**
     * Deletes the node with the given key.
     *
     * <p>Time: O(log n) average
     *
     * @param key the integer key to remove
     * @return {@code true} if the key was found and removed, {@code false} otherwise
     */
    public boolean delete(int key) {
        SkipListNode<T>[] update = buildUpdate(key);

        SkipListNode<T> curr = update[0].forward[0];

        if (curr == null || curr.key != key) {
            return false; // Key not found.
        }

        // Unlink the node from each level.
        for (int i = 0; i <= currentLevel; i++) {
            if (update[i].forward[i] != curr) {
                break; // Node doesn't reach this level.
            }
            update[i].forward[i] = curr.forward[i];
        }

        // Lower currentLevel if top levels became empty.
        while (currentLevel > 0 && head.forward[currentLevel] == null) {
            currentLevel--;
        }

        size--;
        return true;
    }

    /**
     * Returns the number of key-value pairs stored.
     *
     * @return size
     */
    public int size() {
        return size;
    }

    /**
     * Returns a multi-line string showing every level of the skip list.
     *
     * <p>Level 0 is the base (all nodes); higher levels contain fewer nodes.
     * Example output:
     * <pre>
     * Level 3: HEAD -> (7:G)
     * Level 2: HEAD -> (3:C) -> (7:G)
     * Level 1: HEAD -> (1:A) -> (3:C) -> (7:G)
     * Level 0: HEAD -> (1:A) -> (3:C) -> (5:E) -> (7:G)
     * </pre>
     *
     * @return multi-line string representation
     */
    @Override
    public String toString() {
        StringBuilder sb = new StringBuilder();
        for (int i = currentLevel; i >= 0; i--) {
            sb.append(String.format("Level %2d: HEAD", i));
            SkipListNode<T> curr = head.forward[i];
            while (curr != null) {
                sb.append(" -> ").append(curr);
                curr = curr.forward[i];
            }
            sb.append('\n');
        }
        return sb.toString();
    }

    // -------------------------------------------------------------------------
    // Demo main
    // -------------------------------------------------------------------------

    /**
     * Demonstrates insert, search, delete, and the multi-level toString.
     *
     * @param args unused
     */
    public static void main(String[] args) {
        SkipList<String> sl = new SkipList<>();

        System.out.println("=== Skip List Demo ===\n");

        // Insert
        int[] keys = {3, 6, 7, 9, 12, 19, 17, 26, 21, 25};
        String[] vals = {"C", "F", "G", "I", "L", "S", "Q", "Z", "U", "Y"};
        for (int i = 0; i < keys.length; i++) {
            sl.insert(keys[i], vals[i]);
            System.out.printf("Inserted (%d, %s)%n", keys[i], vals[i]);
        }

        System.out.println("\nSkip list after insertions:");
        System.out.println(sl);

        // Search
        System.out.println("Search  9 -> " + sl.search(9));
        System.out.println("Search 17 -> " + sl.search(17));
        System.out.println("Search 99 -> " + sl.search(99));

        // Update
        sl.insert(9, "IX");
        System.out.println("\nAfter updating key 9 to 'IX':");
        System.out.println(sl);

        // Delete
        System.out.println("Delete 12 -> " + sl.delete(12));
        System.out.println("Delete 99 -> " + sl.delete(99));
        System.out.println("\nSkip list after deleting 12:");
        System.out.println(sl);

        System.out.println("Size: " + sl.size());
    }
}

File: python/new_ds/skip_list.py
"""
Skip List
=========
A probabilistic data structure that allows O(log n) average-case search,
insertion, and deletion by maintaining multiple layers of linked lists.

Structure
---------
- Level 0 is the base linked list containing all elements in sorted order.
- Each higher level is a "fast lane" that skips over increasingly many elements.
- A new node is promoted to level k with probability p^k (geometric distribution).

Time Complexity (average case, with p=0.5)
-------------------------------------------
| Operation | Average   | Worst    |
|-----------|-----------|----------|
| search    | O(log n)  | O(n)     |
| insert    | O(log n)  | O(n)     |
| delete    | O(log n)  | O(n)     |

Space Complexity
----------------
- O(n) average (each node appears in ~1/(1-p) levels on average)
- O(n log n) worst case

Why O(log n) average?
---------------------
With p=0.5, the expected number of nodes examined at each level is 2,
and there are O(log n) levels, giving O(log n) total comparisons.
"""

import random
from typing import Any, List, Optional

MAX_LEVEL = 16  # maximum number of levels in the skip list
P = 0.5         # probability of promoting a node to the next level


class SkipListNode:
    """
    A node in the Skip List.

    Attributes
    ----------
    key   : Comparable key used for ordering.
    val   : Associated value stored at this node.
    forward : List of forward pointers, one per level this node occupies.
              forward[0] is the next node at level 0 (base list).
    """

    def __init__(self, key: Any, val: Any, level: int) -> None:
        self.key = key
        self.val = val
        # forward[i] points to the next node at level i; None means end of level
        self.forward: List[Optional["SkipListNode"]] = [None] * (level + 1)

    def __repr__(self) -> str:
        return f"SkipListNode(key={self.key!r}, val={self.val!r}, levels={len(self.forward)})"


class SkipList:
    """
    Skip List with randomized level assignment.

    Maintains keys in sorted order across multiple levels of linked lists.
    The head sentinel node spans all MAX_LEVEL levels.

    Parameters
    ----------
    max_level : int  Maximum number of levels (default MAX_LEVEL=16).
    p         : float  Probability of level promotion (default P=0.5).

    Example
    -------
    >>> sl = SkipList()
    >>> sl.insert(3, "three")
    >>> sl.insert(1, "one")
    >>> sl.search(3)
    'three'
    >>> sl.delete(1)
    >>> sl.search(1) is None
    True
    """

    def __init__(self, max_level: int = MAX_LEVEL, p: float = P) -> None:
        self.max_level = max_level
        self.p = p
        self.current_level = 0  # highest level currently in use (0-indexed)
        self.size = 0
        # Sentinel head node with -inf key; never holds real data
        self._head = SkipListNode(key=float("-inf"), val=None, level=max_level)

    # ------------------------------------------------------------------
    # Internal helpers
    # ------------------------------------------------------------------

    def _random_level(self) -> int:
        """
        Generate a random level for a new node.

        Flips a coin (probability p) repeatedly; the number of heads
        before the first tail determines the level (capped at max_level).

        Time: O(log n) expected
        """
        level = 0
        while random.random() < self.p and level < self.max_level:
            level += 1
        return level

    def _find_update_positions(self, key: Any) -> List[Optional[SkipListNode]]:
        """
        Walk down from the highest level, collecting the last node at each
        level whose key is strictly less than `key`.

        These are the nodes whose forward pointers must be updated during
        insert or delete.

        Returns
        -------
        update : list of length (max_level + 1) where update[i] is the
                 rightmost node at level i with node.key < key.
        """
        update: List[Optional[SkipListNode]] = [None] * (self.max_level + 1)
        current = self._head
        for i in range(self.current_level, -1, -1):
            while current.forward[i] is not None and current.forward[i].key < key:
                current = current.forward[i]
            update[i] = current
        return update

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def insert(self, key: Any, val: Any) -> None:
        """
        Insert or update a key-value pair.

        If `key` already exists its value is overwritten; otherwise a new
        node is created with a randomly assigned level.

        Time:  O(log n) average
        Space: O(log n) for the new node's forward array
        """
        update = self._find_update_positions(key)
        # Check if key already exists at level 0
        candidate = update[0].forward[0] if update[0] is not None else None
        if candidate is not None and candidate.key == key:
            candidate.val = val  # update existing
            return

        new_level = self._random_level()
        # If new node exceeds current height, extend update for new levels
        if new_level > self.current_level:
            for i in range(self.current_level + 1, new_level + 1):
                update[i] = self._head
            self.current_level = new_level

        new_node = SkipListNode(key=key, val=val, level=new_level)
        for i in range(new_level + 1):
            new_node.forward[i] = update[i].forward[i]
            update[i].forward[i] = new_node

        self.size += 1

    def search(self, key: Any) -> Optional[Any]:
        """
        Search for a key and return its associated value, or None if absent.

        Time: O(log n) average
        """
        current = self._head
        for i in range(self.current_level, -1, -1):
            while current.forward[i] is not None and current.forward[i].key < key:
                current = current.forward[i]
        # One step forward at level 0 lands on the candidate
        current = current.forward[0]
        if current is not None and current.key == key:
            return current.val
        return None

    def delete(self, key: Any) -> bool:
        """
        Delete the node with the given key.

        Returns True if the node was found and deleted, False otherwise.

        Time: O(log n) average
        """
        update = self._find_update_positions(key)
        target = update[0].forward[0] if update[0] is not None else None
        if target is None or target.key != key:
            return False  # key not present

        # Splice out the target node at every level it appears
        for i in range(self.current_level + 1):
            if update[i].forward[i] is not target:
                break
            update[i].forward[i] = target.forward[i]

        # Shrink current_level if the top levels are now empty
        while self.current_level > 0 and self._head.forward[self.current_level] is None:
            self.current_level -= 1

        self.size -= 1
        return True

    def __len__(self) -> int:
        return self.size

    def __contains__(self, key: Any) -> bool:
        return self.search(key) is not None

    def __str__(self) -> str:
        """
        Display all levels of the Skip List from top to bottom.

        Example output:
            Level 3: -inf --> 3 --> 7
            Level 2: -inf --> 1 --> 3 --> 7
            Level 1: -inf --> 1 --> 2 --> 3 --> 5 --> 7
            Level 0: -inf --> 1 --> 2 --> 3 --> 5 --> 7
        """
        lines = []
        for level in range(self.current_level, -1, -1):
            nodes = []
            current = self._head
            while current is not None:
                label = "-inf" if current is self._head else str(current.key)
                nodes.append(label)
                current = current.forward[level]
            lines.append(f"Level {level}: " + " --> ".join(nodes))
        return "\n".join(lines)


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

if __name__ == "__main__":
    import time

    print("=" * 60)
    print("SKIP LIST DEMO")
    print("=" * 60)

    sl = SkipList()

    # --- Basic insert and search ---
    data = [(5, "five"), (1, "one"), (9, "nine"), (3, "three"),
            (7, "seven"), (2, "two"), (8, "eight"), (4, "four"), (6, "six")]
    for k, v in data:
        sl.insert(k, v)

    print("\nAll levels after inserting 1..9:")
    print(sl)

    print(f"\nsearch(5)  -> {sl.search(5)!r}")
    print(f"search(10) -> {sl.search(10)!r}  (not present)")

    # --- Update existing key ---
    sl.insert(5, "FIVE_UPDATED")
    print(f"After update: search(5) -> {sl.search(5)!r}")

    # --- Delete ---
    sl.delete(3)
    sl.delete(7)
    print(f"\nAfter deleting 3 and 7:")
    print(sl)
    print(f"search(3) -> {sl.search(3)!r}  (deleted)")

    # --- O(log n) average search demonstration ---
    print("\n--- O(log n) average search benchmark ---")
    big_sl = SkipList()
    N = 100_000
    keys = list(range(N))
    random.shuffle(keys)
    for k in keys:
        big_sl.insert(k, k * 2)

    # Time a batch of searches
    search_keys = random.sample(range(N), 1000)
    start = time.perf_counter()
    for k in search_keys:
        big_sl.search(k)
    elapsed = time.perf_counter() - start
    print(f"1 000 searches in a Skip List of {N:,} elements: {elapsed*1000:.2f} ms")
    print(f"Average per search: {elapsed/1000*1e6:.2f} µs  (expected O(log {N}) ≈ {N.bit_length()} comparisons)")

    print("\nDone.")
